AX: Implement the AXSearchManager class.

AndresGonzalezApple opened the pull request:
#### 8e9119ac64e171d24013c5900c8db27aa9a86cad
<pre>
AX: Implement the AXSearchManager class.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276684">https://bugs.webkit.org/show_bug.cgi?id=276684</a>
&lt;<a href="https://rdar.apple.com/problem/131867309">rdar://problem/131867309</a>&gt;

Reviewed by Tyler Wilcock.

This new class encapsulates the findMatchingObject functionality that was previously in the Accessibility namespace. In addition, it allows to keep state for ranges within objects which is necessary for navigating by misspellings which clients like VoiceOver use. This new functionality will be further used in a subsequent patch. Some code cleanup.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AccessibilitySearchCriteria::AccessibilitySearchCriteria): Deleted.
* Source/WebCore/accessibility/AXLogger.cpp:
* Source/WebCore/accessibility/AXSearchManager.cpp: Added.
(WebCore::isRadioButtonInDifferentAdhocGroup):
(WebCore::AXSearchManager::matchForSearchKeyAtIndex):
(WebCore::AXSearchManager::match):
(WebCore::AXSearchManager::matchText):
(WebCore::AXSearchManager::matchWithResultsLimit):
(WebCore::appendAccessibilityObject):
(WebCore::appendChildrenToArray):
(WebCore::AXSearchManager::findMatchingObjects):
* Source/WebCore/accessibility/AXSearchManager.h: Added.
(WebCore::AccessibilitySearchCriteria::AccessibilitySearchCriteria):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::spellCheckerResults const):
(WebCore::AccessibilityObject::addChild):
(WebCore::AccessibilityObject::findMatchingObjects):
(WebCore::AccessibilityObject::hasMisspelling const): Deleted.
(WebCore::appendAccessibilityObject): Deleted.
(WebCore::Accessibility::isRadioButtonInDifferentAdhocGroup): Deleted.
(WebCore::Accessibility::isAccessibilityObjectSearchMatchAtIndex): Deleted.
(WebCore::Accessibility::isAccessibilityObjectSearchMatch): Deleted.
(WebCore::Accessibility::isAccessibilityTextSearchMatch): Deleted.
(WebCore::Accessibility::objectMatchesSearchCriteriaWithResultLimit): Deleted.
(WebCore::Accessibility::appendChildrenToArray): Deleted.
(WebCore::Accessibility::findMatchingObjects): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityFindMatchingObjects:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::findMatchingObjects):
(WebCore::AXIsolatedObject::spellCheckerResults const):
(WebCore::AXIsolatedObject::hasMisspelling const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: <a href="https://commits.webkit.org/281058@main">https://commits.webkit.org/281058@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7e994d2fbae14906c85dbd91f6b82f928d0a21d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37875 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62174 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8992 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60677 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45511 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9190 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47397 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6406 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60579 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35452 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50622 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28249 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32201 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7924 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7996 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54153 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8196 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63877 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2461 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8204 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54718 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2469 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50648 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54798 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12932 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2080 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33704 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34790 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34535 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1795,11 +1792,12 @@ bool AXIsolatedObject::isSelectedOptionActive() const
+    return Accessibility::retrieveValueFromMainThread<Vector<TextCheckingResult>>([this] () -> Vector<TextCheckingResult> {
+        if (auto* axObject = associatedAXObject())
+            return axObject->spellCheckerResults();
+        return { };
```
We're returning a `Vector<TextCheckingResult>` from the main thread to the secondary thread. Is `TextCheckingResult` and all types within thread-safe?

AndresGonzalezApple replied:
Changed it to returning and storing only the ranges instead of the whole TextSpellingResults. We should be ok as long as the ranges are not accessed in both threads after being created.

twilco reviewed (commented):
I think your commit is missing `AXSearchManager.h` and `AXSearchManager.cpp`.

AndresGonzalezApple committed:
AX: Implement the AXSearchManager class.
https://bugs.webkit.org/show_bug.cgi?id=276684
<rdar://problem/131867309>

Reviewed by Tyler Wilcock.

This new class encapsulates the findMatchingObject functionality that was previously in the Accessibility namespace. In addition, it allows to keep state for ranges within objects which is necessary for navigating by misspellings which clients like VoiceOver use. This new functionality will be further used in a subsequent patch. Some code cleanup.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AccessibilitySearchCriteria::AccessibilitySearchCriteria): Deleted.
* Source/WebCore/accessibility/AXLogger.cpp:
* Source/WebCore/accessibility/AXSearchManager.cpp: Added.
(WebCore::isRadioButtonInDifferentAdhocGroup):
(WebCore::AXSearchManager::matchForSearchKeyAtIndex):
(WebCore::AXSearchManager::match):
(WebCore::AXSearchManager::matchText):
(WebCore::AXSearchManager::matchWithResultsLimit):
(WebCore::appendAccessibilityObject):
(WebCore::appendChildrenToArray):
(WebCore::AXSearchManager::findMatchingObjects):
* Source/WebCore/accessibility/AXSearchManager.h: Added.
(WebCore::AccessibilitySearchCriteria::AccessibilitySearchCriteria):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::spellCheckerResults const):
(WebCore::AccessibilityObject::addChild):
(WebCore::AccessibilityObject::findMatchingObjects):
(WebCore::AccessibilityObject::hasMisspelling const): Deleted.
(WebCore::appendAccessibilityObject): Deleted.
(WebCore::Accessibility::isRadioButtonInDifferentAdhocGroup): Deleted.
(WebCore::Accessibility::isAccessibilityObjectSearchMatchAtIndex): Deleted.
(WebCore::Accessibility::isAccessibilityObjectSearchMatch): Deleted.
(WebCore::Accessibility::isAccessibilityTextSearchMatch): Deleted.
(WebCore::Accessibility::objectMatchesSearchCriteriaWithResultLimit): Deleted.
(WebCore::Accessibility::appendChildrenToArray): Deleted.
(WebCore::Accessibility::findMatchingObjects): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityFindMatchingObjects:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::findMatchingObjects):
(WebCore::AXIsolatedObject::spellCheckerResults const):
(WebCore::AXIsolatedObject::hasMisspelling const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: https://commits.webkit.org/281058@main
